activesupport/test/descendants_tracker_test_cases.rb

```diff
@@ -28,11 +28,23 @@ def test_descendants
   end
 
   def test_descendants_with_garbage_collected_classes
-    1.times do
+    # The Ruby GC (and most other GCs for that matter) are not fully precise.
+    # When GC is run, the whole stack is scanned to mark any object reference
```

Comment:
The whole _heap_ is scanned, right? I don't think Ruby has any concept of stack-allocated values. In which case I don't understand this explanation or the fix; threads don't have separate heaps.

Reply:
The heap is scanned yes, but the stack too.

>  I don't think Ruby has any concept of stack-allocated values.

No, but the stack contain references to heap objects (`VALUE`).

> threads don't have separate heaps.

But they have a separate stack.

Reply:
Here's the code that mark the stack: https://github.com/ruby/ruby/blob/fc78a0890481e092ec9737c1ab3cbd4c32793eaa/gc.c#L6413-L6421

Reply:
You're right, of course, the GC has to know whether objects are still referenced from the stack. If `1.times` was creating a stack frame though, surely the anonymous class was being marked already? I wonder if forcing a full GC run is enough to fix the problem, and the thread was just making it more likely that the object would be GC'd anyway.

Reply:
I believe the `1.times` was just to create a new lexical scope for `child_klass`, so that the reference would be out of scope when `GC.start` was called.

Reply:
> If `1.times` was creating a stack frame though, surely the anonymous class was being marked already?

No that's the thing. It's a bit tricky to explain, but I'll try. I ran into a very similar issue a while ago, with objects that wouldn't be GCed, and a colleague, Alan Wu (or was it Aaron Patternson?) investigated it and explained it to me (so that's kind of second hand knowledge). 

So let's say we have 4 registers, `a`, `b`, `c` and `d`.

Somehow when we do `1.times { o = Class.new(Parent)}`, a reference to that object is written in `d`, and it stay there because nothing else uses that register during the frame, and the frame is pushed.

When we exit that frame the register is restored, and then we call another function (e.g. `GC.start`) that simply don't use register `d`, that register is saved again, even though it's useless. 

Now when GC mark the stack, it can't tell that this register content was a dead reference, so the object is considered still referenced.

Also maybe to clarify, I'm talking about the C stack here, with actual physical registers.

Reply:
Okay, this is far past the limit of what I know about Ruby internals, so I'm happy to go with the fix as is. Thank you for the explanation!

Reply:
I'm still not entirely clear why the failure started showing up after #39428, though.

Incidentally, I cannot reproduce the failure locally with only without running `test/current_attributes_test.rb`, and when I first discovered [the failure in another PR](https://buildkite.com/rails/rails/builds/79536), I rebased my code to before 8195cd5f9970960cbd25ba1383f893a977014c81, and [it passed](https://buildkite.com/rails/rails/builds/79537).

Reply:
I suppose it's possible that there are actually two failures involved here: (1) the GC issues discussed above (which I'm not seeing locally), and (2) some interaction between GC and thread locals caused by the `Enumerator`s in #39428 (which I can repro locally).  And wrapping in `Thread.new` mitigates them both. :thinking:

Reply:
> I'm still not entirely clear why the failure started showing up after #39428, though.

Yeah, me neither.